source/extensions/stat_sinks/common/statsd/statsd.cc

```diff
@@ -38,34 +40,58 @@ void UdpStatsdSink::WriterImpl::write(const std::string& message) {
   Network::Utility::writeToSocket(*io_handle_, &slice, 1, nullptr, *parent_.server_address_);
 }
 
+void UdpStatsdSink::WriterImpl::writeBuffer(Buffer::Instance* data) {
+  Network::Utility::writeToSocket(*io_handle_, *data, nullptr, *parent_.server_address_);
+}
+
 UdpStatsdSink::UdpStatsdSink(ThreadLocal::SlotAllocator& tls,
                              Network::Address::InstanceConstSharedPtr address, const bool use_tag,
-                             const std::string& prefix)
+                             const std::string& prefix, uint64_t buffer_size)
     : tls_(tls.allocateSlot()), server_address_(std::move(address)), use_tag_(use_tag),
-      prefix_(prefix.empty() ? Statsd::getDefaultPrefix() : prefix) {
+      prefix_(prefix.empty() ? Statsd::getDefaultPrefix() : prefix), buffer_size_(buffer_size) {
   tls_->set([this](Event::Dispatcher&) -> ThreadLocal::ThreadLocalObjectSharedPtr {
     return std::make_shared<WriterImpl>(*this);
   });
 }
 
 void UdpStatsdSink::flush(Stats::MetricSnapshot& snapshot) {
-  Writer& writer = tls_->getTyped<Writer>();
+  auto buffer = Buffer::OwnedImpl();
   for (const auto& counter : snapshot.counters()) {
     if (counter.counter_.get().used()) {
-      writer.write(absl::StrCat(prefix_, ".", getName(counter.counter_.get()), ":", counter.delta_,
-                                "|c", buildTagStr(counter.counter_.get().tags())));
+      auto counterStr =
+          absl::StrCat(prefix_, ".", getName(counter.counter_.get()), ":", counter.delta_, "|c",
+                       buildTagStr(counter.counter_.get().tags()));
+      if (buffer.length() + counterStr.length() > buffer_size_) {
+        flushBuffer(buffer);
+      }
+      buffer.add(counterStr);
```

Comment:
is there a concern that counterStr may be by itself larger than buffer_size_?

I don't remember the exact semantics, but note that you are guarding against overflowing the buffer above but then possibly overflowing it on the add().

Consider `counter_str.size() > buffer_size_` and buffer.length() == 0; you will call `flushBuffer` here on an empty buffer and then append a string that's too large.

Reply:
Hey @jmarantz, thanks for the review!

I've updated the code to properly check the size of the data being added to the buffer to the `buffer_size_` value. Currently however the buffer being used is per invocation of `UdpStatsdSink::flush` and isn't reserving space before operating. I'm not sure what a good pattern here would be or if there are some standard usage patterns I should be aware of (guidance definitely appreciated 😄 ).

Reply:
Oh, also added a few tests that were sorely needed to exercise a few conditions.

0. The value added is larger than the buffer
0. Multiple values added less than the buffer maximum
0. Multiple values added exceeding the buffer size

Let me know if there are any other cases that might be worth testing.